scanner/decisions/0010-separate-versioning-for-vuln-bundles.md

```diff
@@ -0,0 +1,84 @@
+
+Initially, two version streams will exist (table 1 below):
+
+| Stream | Git Ref      | Description                                                        |
+|--------|--------------|--------------------------------------------------------------------|
+| dev    | heads/master | This stream tracks the latest vulnerability schema                 |
+| v1     | heads/4.4.5  | This stream tracks the schema used by the current StackRox release |
+
+On the subsequent releases (for example,. `4.4.6`), one of the two will be carried out:
+
+1.  The schema hasn't changed, so v1 now tracks the new release (`heads/4.4.6`).
+2.  The schema has changed, so v1 stays at the current release (`heads/4.4.5`), and a new version, v2, is created to track the next release (`heads/4.4.6`).
+
+### Building versioned bundles
+
+The GitHub workflow will have a map with "bundle version" as keys and "git reference" as values, similar to Table 1 above.  This can be stored statically in the matrix definition.
+
+For each version in the mapping, a bundle will be built based on the specified git ref and stored at `https://definitions.stackrox.io/v4/<bundle-version>/vulnerabilities.zip`.
+
+In every release cut, the key specified in `scanner/VULNERABILITY_VERSION` will be updated in the mapping to the current release branch reference (e.g., `heads/4.4.6`).  Notice that if the version is the same, it will be updated.  If it's new, it will be added, and the previous version will stay.
+
+Offline bundles per release will not be created anymore, but per bundle version.  The latest offline bundle will ship all bundle versions currently supported by StackRox.
+
+### Ingesting versioned bundles
+
+Scanner will know the vulnerability version it requires based on hard-coded information added during build time.  A file at `scanner/VULNERABILITY_VERSION` will determine the version on release builds and nightlies.  On non-release builds this will hard-coded to "dev".
+
+The definitions handler in Central can be kept the same.  It already supports arbitrary versions in the GET request from Scanner.  [The check for nightlies](https://github.com/stackrox/stackrox/blob/e332cd19a639f59d9931414a4f4e561981396dad/central/scannerdefinitions/handler/handler.go#L661) and other per-release checks in the ingestion path are unnecessary and will be removed.
+
+The bundle version will be configured in the vulnerability URL by a new variable called `ROX_VULNERABILITY_VERSION` to specify the expected bundle version to fetch.  The previous variable `ROX_VERSION` will be left for backward compatibility.
+
+The vulnreability bundle will include a `manifest.json` file, which will carry the StackRox version and bundle version used to build it.  This file will be used by Scanner to prevent ingestion of unsupported bundle versions.
+
+In offline mode, the `roxctl` method for fetching bundles will be modified to fetch only the latest bundle.
```

Comment:
Not a fan of this, would recommend keeping the logic to choose the bundle (v1, v2, 4.4, etc.) that has data appropriate to current ACS version. 

The latest bundle is currently 340MB, adding this new `v1` will make that 500MB, if we 'fix' the 4.4 issue that will be 670MB, etc..  IMHO we should focus on making the bundles minimal to reduce costs and improve the user experience (the larger the file to manage the more likely will be issues with storage, transfers, wait times, etc.)

Reply:
Oh, so it's even smaller than 1GiB. I think we can go with not supporting 4.4 (not GA), and we can ship 4.5 + v1 (~500MiB) until we have v2 (where we can potentially remove 4.5? Or do the hacky thing I mentioned before? to keep the bundle around <600MiB). Is it really a pain to have a bundle of that size for the benefit of having one offline bundle for all releases?

Reply:
> I think we can go with not supporting 4.4 (not GA), and we can ship 4.5 + v1 (~500MiB) until we have v2 (where we can potentially remove 4.5?

I'm skeptical if we'll be able to remove 4.5 - I'd bet on having to keep it around 'years' if current behaviors are an indicator.

> Is it really a pain to have a bundle of that size for the benefit of having one offline bundle for all releases?

It may be, we may even run into limits in the `emptyDir` in the Central deploy.  If there is consensus for the super bundle so be it, however my vote / recommendation is to keep the version specific bundles (`v1`, `v2`, `4.5`, etc.) as the long term strat (should be simple enough as we'll only have a few right?)  Another thought is on download/upload (via roxctl) we can 'repackage' the zip so that we only upload the specific pieces needed.

Still if size becomes a problem (slow pipe?) then a file for the specific version will be more efficient overall.

Perhaps another thing to consider is we eventually will need scanner-v4-indexer/matcher to return a version to report status (ROX-21362) - be simple to add the 'VULNERABILITY_VERSION' to the same response (and get a head start)

Reply:
I think the path to versioned offline bundles are very possible, and not ambiguous -- it just require additional work. Like I said above, right now I am concerned only with the size impact more than anything -- if the size increase in the latest bundle is indeed painful to the point it make the flow worse for customers compared to migrating to roxctl in their scripts, then I would change my opinion on moving forward with the versioned offline bundles.

Reply:
Perhaps another consideration - entertaining the scenario a user DOES have an issue with a 600MB+ offline bundle, what would our options be to help that user?  

With minimal versioned bundles we have an option - use the versioned bundle.

With the giant bundle, assuming downloads aren't an issue, we would ask the user to download the bundle, extract it, remove ~50%+ of the data that is unused, and then try the upload again - the user would have to do this every time they download the offline bundle (daily+)

Reply:
It was determined that offline bundle size is a concern for certain use-cases. Then, the decision will support both paths: versioned and latest offline bundles, with `roxctl` being able to pull offline bundles using `download-db` the same as today.